File: apps/web/src/lib/terminal-protocol.ts
export type TerminalClientMessage =
  | { type: "input"; data: string }
  | { type: "resize"; cols: number; rows: number }
  | { type: "ping" };

export type TerminalServerMessage =
  | { type: "ready"; cols: number; rows: number }
  | { type: "output"; data: string }
  | {
      type: "status";
      phase: "starting" | "running" | "stopped";
      detail?: string;
    }
  | { type: "exit"; exitCode: number | null }
  | { type: "error"; message: string };

File: services/runner/internal/engine/terminal_protocol.go
package engine

const TerminalCommandExitMarker = "__GITGYM_COMMAND_EXIT__"

type TerminalClientMessage struct {
	Type string `json:"type"`
	Data string `json:"data,omitempty"`
	Cols uint16 `json:"cols,omitempty"`
	Rows uint16 `json:"rows,omitempty"`
}

type TerminalServerMessage struct {
	Type     string `json:"type"`
	Data     string `json:"data,omitempty"`
	Phase    string `json:"phase,omitempty"`
	Detail   string `json:"detail,omitempty"`
	Cols     uint16 `json:"cols,omitempty"`
	Rows     uint16 `json:"rows,omitempty"`
	ExitCode *int   `json:"exitCode,omitempty"`
	Message  string `json:"message,omitempty"`
}
